### PR TITLE
Fix/#414 Formatting version list with colors on supported terminal only

### PR DESCRIPTION
**Pull request by trgiangdo**
This PR adds `supports_color()` method to check if the terminal supports colorization, and some other minor fixes on the formatting.

Remarks:
- Works fine on color supported terminal.
- Works fine in notebook environment.

### CODE REVIEW

File: src/taipy/core/_version/bcolor.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        try:
+            reg_key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, "Console")
+            reg_key_value, _ = winreg.QueryValueEx(reg_key, "VirtualTerminalLevel")
+        except FileNotFoundError:
+            return False
+        else:
+            return reg_key_value == 1
+
+
+def supports_color():
```
Comment (jrobinAV) [inline src/taipy/core/_version/bcolor.py]:
```suggestion
def _is_color_supported():
```

File: src/taipy/core/_version/bcolor.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+
+class Bcolors:
+    PURPLE = "\033[95m" if supports_color() else ""
+    BLUE = "\033[94m" if supports_color() else ""
+    CYAN = "\033[96m" if supports_color() else ""
+    GREEN = "\033[92m" if supports_color() else ""
+    BOLD = "\033[1m" if supports_color() else ""
+    UNDERLINE = "\033[4m" if supports_color() else ""
+    END = "\033[0m" if supports_color() else ""
```
Comment (jrobinAV) [inline src/taipy/core/_version/bcolor.py]:
```suggestion
    PURPLE = "\033[95m" if _is_color_supported() else ""
    BLUE = "\033[94m" if _is_color_supported() else ""
    CYAN = "\033[96m" if _is_color_supported() else ""
    GREEN = "\033[92m" if _is_color_supported() else ""
    BOLD = "\033[1m" if _is_color_supported() else ""
    UNDERLINE = "\033[4m" if _is_color_supported() else ""
    END = "\033[0m" if _is_color_supported() else ""
```

File: src/taipy/core/_version/bcolor.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    return is_a_tty and (
+        sys.platform != "win32"
+        or "ANSICON" in os.environ
+        or "WT_SESSION" in os.environ  # Windows Terminal supports VT codes.
+        or os.environ.get("TERM_PROGRAM") == "vscode"  # VSCode's built-in terminal supports colors.
+        or vt_codes_enabled_in_windows_registry()
+    )
+
+
+class Bcolors:
```
Comment (jrobinAV) [inline src/taipy/core/_version/bcolor.py]:
```suggestion
class _Bcolors:
```

File: src/taipy/core/_version/_version_cli.py [hunk 4]
```diff
@@ -109,24 +101,30 @@ def _parse_arguments(cls):
             versions = _VersionManagerFactory._build_manager()._get_all()
             versions.sort(key=lambda x: x.creation_date, reverse=True)
 
+            bold = False
             for version in versions:
                 if version.id == development_version_number:
-                    list_version_message += bcolors.GREEN
+                    list_version_message += Bcolors.GREEN
```
Comment (jrobinAV) [inline src/taipy/core/_version/_version_cli.py]:
```suggestion
                    list_version_message += _Bcolors.GREEN
```

File: src/taipy/core/_version/_version_cli.py [hunk 5]
```diff
@@ -109,24 +101,30 @@ def _parse_arguments(cls):
...
                     mode = "Development"
                 elif version.id in production_version_numbers:
-                    list_version_message += bcolors.PURPLE
+                    list_version_message += Bcolors.PURPLE
```
Comment (jrobinAV) [inline src/taipy/core/_version/_version_cli.py]:
```suggestion
                    list_version_message += _Bcolors.PURPLE
```

File: src/taipy/core/_version/_version_cli.py [hunk 6]
```diff
@@ -109,24 +101,30 @@ def _parse_arguments(cls):
...
                     mode = "Production"
                 else:
-                    list_version_message += bcolors.BLUE
+                    list_version_message += Bcolors.BLUE
```
Comment (jrobinAV) [inline src/taipy/core/_version/_version_cli.py]:
```suggestion
                    list_version_message += _Bcolors.BLUE
```

File: src/taipy/core/_version/_version_cli.py [hunk 7]
```diff
@@ -109,24 +101,30 @@ def _parse_arguments(cls):
...
                     mode = "Experiment"
 
                 if version.id == latest_version_number:
-                    list_version_message += bcolors.BOLD
+                    list_version_message += Bcolors.BOLD
```
Comment (jrobinAV) [inline src/taipy/core/_version/_version_cli.py]:
```suggestion
                    list_version_message += _Bcolors.BOLD
```

File: src/taipy/core/_version/_version_cli.py [hunk 8]
```diff
@@ -109,24 +101,30 @@ def _parse_arguments(cls):
...
+                    bold = True
                     mode += " (latest)"
 
                 list_version_message += (
                     f"{(version.id):<36}   {mode:<20}   {version.creation_date.strftime('%Y-%m-%d %H:%M:%S'):<20}\n"
                 )
+                list_version_message += Bcolors.END
```
Comment (jrobinAV) [inline src/taipy/core/_version/_version_cli.py]:
```suggestion
                list_version_message += _Bcolors.END
```

File: src/taipy/core/_version/_version_cli.py [hunk 9]
```diff
@@ -109,24 +101,30 @@ def _parse_arguments(cls):
...
+
+                if bold:
+                    list_version_message += Bcolors.END
```
Comment (jrobinAV) [inline src/taipy/core/_version/_version_cli.py]:
```suggestion
                    list_version_message += _Bcolors.END
```

File: src/taipy/core/_version/_version_cli.py [hunk 10]
```diff
@@ -14,15 +14,7 @@
...
-
-
-class bcolors:
-    PURPLE = "\033[95m"
-    BLUE = "\033[94m"
-    CYAN = "\033[96m"
-    GREEN = "\033[92m"
-    BOLD = "\033[1m"
-    UNDERLINE = "\033[4m"
+from .bcolor import Bcolors
```
Comment (jrobinAV) [inline src/taipy/core/_version/_version_cli.py]:
```suggestion
from .bcolor import _Bcolors
```

File: src/taipy/core/_version/bcolor.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+
+import os
+import sys
+
+
+def vt_codes_enabled_in_windows_registry():
```
Comment (jrobinAV) [inline src/taipy/core/_version/bcolor.py]:
```suggestion
def _vt_codes_enabled_in_windows_registry():
```

File: src/taipy/core/_version/bcolor.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    """
+
+    is_a_tty = hasattr(sys.stdout, "isatty") and sys.stdout.isatty()
+
+    return is_a_tty and (
+        sys.platform != "win32"
+        or "ANSICON" in os.environ
+        or "WT_SESSION" in os.environ  # Windows Terminal supports VT codes.
+        or os.environ.get("TERM_PROGRAM") == "vscode"  # VSCode's built-in terminal supports colors.
+        or vt_codes_enabled_in_windows_registry()
```
Comment (jrobinAV) [inline src/taipy/core/_version/bcolor.py]:
```suggestion
        or _vt_codes_enabled_in_windows_registry()
```

**Review by jrobinAV — COMMENTED** [review]
Some naming remarks.

**Review by jrobinAV — APPROVED** [review]

**Comment by trgiangdo** [comment]
About the unit test if the terminal supports color or not, first of all, I don't have a specific way to test this now for unsupported terminal.

In addition, now rethink about it, it is really pointless imo. It is based on which terminal we run it. If mock the return value of `_is_color_supported()` to False, then we can't know for sure if it will work on a specific terminal.

**Comment by trgiangdo** [comment]
On why the coverage of `_bcolor.py` is so low, it is because of a function that check for windows registry.

IMO, I don't think we need to test this (since it is a hacky way to print with color on unsupported color terminal).

If we need to, let me know and I may find a way. There is no clear solution at the moment.

**Review by jrobinAV — APPROVED** [review]
I agree.

**Review by jrobinAV — APPROVED** [review]

**Pull request merged**